skytemple_files/patch/handler/ai_item_improvements.py

```diff
@@ -0,0 +1,200 @@
+from skytemple_files.common.ppmdu_config.data import (
+    Pmd2Data,
+    GAME_VERSION_EOS,
+    GAME_REGION_US,
+    GAME_REGION_EU,
+)
+from skytemple_files.patch.category import PatchCategory
+from skytemple_files.patch.handler.abstract import AbstractPatchHandler
+from skytemple_files.common.i18n_util import _
+
+MODIFIED_INSTRUCTION = 0xE8BD83F8  # ldmia all the shit
+OFFSET_EU = 0x231F29C - 0x22DCB80  # location of the instruction minus starting point of overlay 29
+OFFSET_US = 0x231E834 - 0x22DC240  # same but US HeHeHaHa
+
+ADD_TYPES_ADDRESS_NA = 0x2EAA4
+ADD_TYPES_ADDRESS_EU = 0x2EBD8
+ADD_TYPES_ORIGINAL_CODE = 0xE3A00024
+
+
+# the function where all the magic happens
+def set_item_ai_flags(item_p_model, item_id: int, ai_flag_1: bool, ai_flag_2: bool, ai_flag_3: bool):
+    item = item_p_model.item_list[item_id]
+    item.ai_flag_1 = ai_flag_1
+    item.ai_flag_2 = ai_flag_2
+    item.ai_flag_3 = ai_flag_3
+
+
+class AiItemImprovementsHandler(AbstractPatchHandler):
+    @property
+    def name(self) -> str:
+        return "AiItemImprovements"
+
+    @property
+    def description(self) -> str:
+        return _("Tweaks team and enemy AI to use more types of items, and use them more effectively. Overwrites AiItemOptimizations.")
+
+    @property
+    def author(self) -> str:
+        return "happylappy"
+
+    @property
+    def version(self) -> str:
+        return "1.0.0"
+
+    @property
+    def category(self) -> PatchCategory:
+        return PatchCategory.IMPROVEMENT_TWEAK
+
+    def is_applied(self, rom: NintendoDSRom, config: Pmd2Data) -> bool:
+        patch_area = get_binary_from_rom(rom, config.bin_sections.overlay29)
+        if config.game_version == GAME_VERSION_EOS:
+            if config.game_region == GAME_REGION_US:
+                return read_u32(patch_area, OFFSET_US) == MODIFIED_INSTRUCTION
+            if config.game_region == GAME_REGION_EU:
+                return read_u32(patch_area, OFFSET_EU) == MODIFIED_INSTRUCTION
+        raise NotImplementedError()
+
+    def apply(self, apply: Callable[[], None], rom: NintendoDSRom, config: Pmd2Data) -> None:
+        param = self.get_parameters()
+        # do an is_applied check for AddTypes. This is necessary because the Gummi table size changes if this patch is applied!
+        if config.game_version == GAME_VERSION_EOS:
+            if config.game_region == GAME_REGION_US:
+                add_types_applied = (
+                    read_u32(rom.loadArm9Overlays([29])[29].data, ADD_TYPES_ADDRESS_NA) != ADD_TYPES_ORIGINAL_CODE
+                )
+            elif config.game_region == GAME_REGION_EU:
+                add_types_applied = (
+                    read_u32(rom.loadArm9Overlays([29])[29].data, ADD_TYPES_ADDRESS_EU) != ADD_TYPES_ORIGINAL_CODE
+                )
+            else:
+                raise NotImplementedError()
+        else:
+            raise NotImplementedError()
+        if add_types_applied
+            param["AddTypesApplied"] = 1
```

Comment:
```suggestion
            param["AddTypesApplied"] = 1 # this DOES violate convention, but the ASM code needs to change if AddTypes is applied (for reasons explained above), and this seems to be the only way to pass information from the handler to the ASM code, other than asking the user in the config, which seems redundant when AddTypesApplied can be determined automatically.
```

Hi hello I wrote a lot of the patch handler code, here's the explanation.

Reply:
Tysm chesyon